keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -0,0 +1,492 @@
+`Pod` object is actually a
+VM running in a hypervisor, and the `PVC` is actually a virtual disk attached to a VM.
+It's common for virtualization systems to allows VMs to boot from disks, and for disks
+to be pre-populated with various OS images. OS images tend to be stored in external
+repositories dedicated to that purpose, often with various mechanisms for retrieving
+them efficiently that are external to Kubernetes.
+
+One way to achieve this is to represent disk images as custom resources that point to
+the image repository, and to allow creation of PVCs from these custom resources such
+that the volumes come pre-populated with the correct data. Efficient population of the
+data could be left up to a purpose-built controller that knows how to get the bits
+where they need to be with minimal I/O.
+
+#### Backup/Restore
+
+Without getting into the details of how backup/restore should be implemented, it's
+clear that whatever design one chooses, a necessary step is to have the user
+(or higher level controller) create a PVC that points to the backup they want to
+restore, and have the data appear in that volume somehow.
+
+One can imagine backups simply being a special case of snapshots, in which case the
+existing design is sufficient, but if you want anything more sophisticated, there
+will inevitably be a new object type that represents a backup. While it's arguable
+that backup should be something CSI plugins should be exclusively responsible for,
+one can also argue that generic backup tools should also exist which can backup
+and restore all kind of volumes. Those tools will be apart from CSI plugins and
+yet need a way to populate volumes with restored volumes.
+
+It's also likely that multiple backup/restore implementations will be developed,
+and it's not a good idea to pick a winner at the Kubernetes API layer. It makes
+more sense to enable developers to try different approaches by making the API allow
+restoring from various kinds of things.
+
+### Implementation Details/Notes/Constraints
+
+As noted above, the proposal is extremely simple -- just remove the validation on
+the `DataSource` field. This raises the question of WHAT will happen when users
+put new things in that field, and HOW populators will actually work with so small
+a change.
+
+It's first important to note that only consumers of the `DataSource` field today
+are the various dynamic provisioners, most notably the external-provisioner CSI
+sidecar. If the external-provisioner sidecar sees a data source it doesn't
+understand, it simply ignores the request, which is both important for forward
+compatibility, and also perfect for the purposes of a data populator. This allows
+developers to add new types of data sources that the dynamic provisioners will
+simply ignore, enabling a different controller to see these objects and respond
+to them.
+
+I will leave the details of how data populators will work for another KEP. There
```

Comment:
I'm a bit uncomfortable with this - I'd like to at lease know that it is possible to implement a trivial populator without another API change.

I assume the flow is something like:

* user makes a pod and PVC requesting a populator
* populator controller runs a new PVC' (with reclaim policy = retain) and a pod' to use it
* real underlying provisioner makes a volume and binds to PVC'
* populator pod' runs and initializes PV
* populator pod' exits
* populator controller deletes the PVC'
* populator modifies the PV to bind to user's PVC
* user's pod can proceed

Something like that?  I'd like to see the details.

Reply:
Yes that's exactly the approach that the prototype uses. I didn't want to make this particular proposal more complicated by including those details because this KEP relates to the in-tree portion of the work, while the populator guts are expected to all live out of tree.

I'm happy to include more detail in this proposal but it means that, as we put the finishing touches on the out-of-tree work, the KEP might need further updates in the future related to that work.

Reply:
Even though the work is out of tree, I think it's still valuable to include here for a complete picture (we've done this for many features that have both in and out of tree components)

Reply:
I just want enough to believe that it can work.